### PR TITLE
fix(provision): clean playwright artifacts during sanity check

### LINKED ISSUE #584: bug: playwright dependencies installation (or tests) leaves a set of unwanted files in the `jenkins` user home directory
While working on https://github.com/jenkins-infra/jenkins-infra/pull/2744, @lemeurherve caught that there was unexpected elements in the `/home/jenkins` user home directory, related to the `playwright` installation (or testing) process:

```
+ ls -la /home/jenkins
# ...
 drwxrwxr-x 195 jenkins jenkins  4096 Apr  3 13:20 node_modules
 drwxrwxr-x   4 jenkins jenkins  4096 Apr  3 13:19 .npm
 -rw-rw-r--   1 jenkins jenkins    60 Apr  3 13:20 package.json
 -rw-rw-r--   1 jenkins jenkins 97457 Apr  3 13:20 package-lock.json
# ...
```

Particularly:

```
+ cat /home/jenkins/package.json
 {
   "dependencies": {
     "playwright-test": "^8.2.0"
   }
 }
```

We should have these elements removed

### PR DESCRIPTION
**Pull request by hainenber**
## What does this PR do?
Fixes #584 

## Description
Remove build artifacts accidentally introduced during `playwright`'s sanity check. The fix is backwards-compatible as it registers the installed `playwright`'s version somewhere before cleaning up all.

### CODE REVIEW

**Review by dduportal — COMMENTED** [review]
Discussion happening in https://github.com/jenkins-infra/packer-images/issues/584

This PR looks good though (It makes me realize that the implementation does not map to what I had in mind and that I described in https://github.com/jenkins-infra/packer-images/issues/584#issuecomment-1712473868 though): we *are* installing `playwright` in 'latest' while we should not

cc @smerle33 to look into it to help our contributor here and select an improvement path (e.g. pinning playwright + tracking it with `updatecli` or only install its dependencies)

**Comment by smerle33** [comment]
> Discussion happening in #584
> 
> This PR looks good though (It makes me realize that the implementation does not map to what I had in mind and that I described in [#584 (comment)](https://github.com/jenkins-infra/packer-images/issues/584#issuecomment-1712473868) though): we _are_ installing `playwright` in 'latest' while we should not
> 
> cc @smerle33 to look into it to help our contributor here and select an improvement path (e.g. pinning playwright + tracking it with `updatecli` or only install its dependencies)

I agree, we should avoid the latest and pin the version like we do with others tools from the file https://github.com/jenkins-infra/packer-images/blob/main/provisioning/tools-versions.yml 
and then use updatecli to track changes.

as for removing the playwright-test and its dependency I would either try to use a npm command to remove it or even, as proposed, remove the sanity check of version (but keep a call to make sure it respond), and move the version check to goss like we did with jdk versions : https://github.com/jenkins-infra/packer-images/blob/main/goss/goss.yaml

WDYT @hainenber  ? we can pair on this or work asynchronously as you wish

**Comment by hainenber** [comment]
Hey there, I think you've made a great outline of what items could be done here. I can work on this async and may need your support for further testing. Let me know if it's good to you. Thanks!

**Comment by hainenber** [comment]
Hi there, I've made amends to (1) install `playwright` with pinned version and (2) set up `updatecli`'s config to automate the dependency management.

One concern from my side is that downstream process(es) may install duplicating `playwright` and resource contention could happen (pods/VMs out of storage, for example). I believe we should remove `playwright` installation before merging this one.

File: goss/goss.yaml [hunk 1]
```diff
@@ -24,6 +24,11 @@ command:
     stdout:
       - 21-beta
     stderr: []
+  playwright --version:
+    exit-status: 0
+    stdout:
+      - 1.37.1
```
Comment (smerle33) [inline goss/goss.yaml:30]: seem not to be the value received : 
```
Command: playwright --version: exit-status:
[2023-09-13T16:25:14.233Z] 2023-09-13T16:25:14Z:     azure-arm.ubuntu: Expected
[2023-09-13T16:25:14.233Z] 2023-09-13T16:25:14Z:     azure-arm.ubuntu:     <int>: 127
[2023-09-13T16:25:14.233Z] 2023-09-13T16:25:14Z:     azure-arm.ubuntu: to equal
[2023-09-13T16:25:14.233Z] 2023-09-13T16:25:14Z:     azure-arm.ubuntu:     <int>: 0
[2023-09-13T16:25:14.233Z] 2023-09-13T16:25:14Z:     azure-arm.ubuntu: Command: playwright --version: stdout: patterns not found: [1.37.1]
```

File: goss/goss.yaml [hunk 2]
```diff
@@ -24,6 +24,11 @@ command:
     stdout:
       - 21-beta
     stderr: []
+  playwright --version:
```
Comment (smerle33) [inline goss/goss.yaml:27]:
```suggestion
  playwright --version:
  /home/jenkins/.asdf/shims/playwright --version:
```

File: goss/goss.yaml (hunk 1 above)
Comment (hainenber) [reply goss/goss.yaml:30]: I've approved your suggested change. Does it work now?

Comment (smerle33) [reply goss/goss.yaml:30]: it mean changing also the updatecli manifest and then the dot at `.asdf` make it difficult for key name ... I suggest to change the goss.yaml to use the new version with exec as here : https://github.com/goss-org/goss/blob/master/docs/manual.md#command

but then we will have to update the updatecli manifest for jdk that are updating that file

Comment (smerle33) [reply goss/goss.yaml:30]: you can try locally with this : https://github.com/jenkins-infra/packer-images/blob/main/README.md#invoking-the-build-locally

File: goss/goss.yaml (hunk 2 above)
Comment (smerle33) [reply goss/goss.yaml]: @hainenber  I probably sent you in the wrong way here, the problem is not the path but the context, let me try a little more and I will show/explain this

**Comment by smerle33** [comment]
> I believe we should remove `playwright` installation in the downstream before merging this one. Hopefully I'm making sense on this one 😅

Yes, I agree, can you start a search on both `jenkins-infra` and `jenkinsci` repositories, for projects that are using playwright ?

**Comment by hervelemeur** [comment]
Playwrigt is used only for jenkins.io component, cf https://github.com/jenkins-infra/helpdesk/issues/3345

File: goss/goss.yaml (hunk 2 above)
Comment (smerle33) [reply goss/goss.yaml]: FYI
```
jenkins@ac02732a066b:/tmp$ cat /home/jenkins/.asdf/shims/playwright 
#!/usr/bin/env bash
# asdf-plugin: nodejs 18.17.0
exec /home/jenkins/.asdf/bin/asdf exec "playwright" "$@" # asdf_allow: ' asdf '
```

**Comment by hainenber** [comment]
> > I believe we should remove `playwright` installation in the downstream before merging this one. Hopefully I'm making sense on this one 😅
> 
> @hainenber Yes, I agree, can you start a search on both `jenkins-infra` and `jenkinsci` repositories, for projects that are using playwright ?

Regarding this, I've had a second thought and believe we should carry the transition in a safer manner by keeping both `playwright` installation before deprecating the downstream one. Perhaps I was overthinking on the storage issue 😄

File: goss/goss.yaml (hunk 2 above)
Comment (hainenber) [reply goss/goss.yaml]: The `exec` command to check for `playwright` version is now ran under `jenkins` user. Successfully built and tested on my local machine

Screenshot 2023-09-21 at 23 43 52

Comment (hainenber) [reply goss/goss.yaml]: The `exec` command to check for `playwright` version is now ran under `jenkins` user. Successfully built and tested on my local machine

Screenshot 2023-09-21 at 23 43 52

**Comment by smerle33** [comment]
Hi,
we still hit errors on builds for VM only (not docker) : 
```
[2023-09-23T10:26:55.531Z] 2023-09-23T10:26:53Z: ==> azure-arm.ubuntu: + goss --version

[2023-09-23T10:26:55.531Z] 2023-09-23T10:26:53Z:     azure-arm.ubuntu: goss version v0.3.20

[2023-09-23T10:26:55.531Z] 2023-09-23T10:26:53Z: ==> azure-arm.ubuntu: + goss --gossfile /tmp/goss.yaml validate --retry-timeout 5s

[2023-09-23T10:26:55.531Z] 2023-09-23T10:26:54Z:     azure-arm.ubuntu: ...............FF

[2023-09-23T10:26:55.531Z] 2023-09-23T10:26:54Z:     azure-arm.ubuntu:

[2023-09-23T10:26:55.531Z] 2023-09-23T10:26:54Z:     azure-arm.ubuntu: Failures/Skipped:

[2023-09-23T10:26:55.531Z] 2023-09-23T10:26:54Z:     azure-arm.ubuntu:

[2023-09-23T10:26:55.531Z] 2023-09-23T10:26:54Z:     azure-arm.ubuntu: Command: playwright_version: exit-status:

[2023-09-23T10:26:55.531Z] 2023-09-23T10:26:54Z:     azure-arm.ubuntu: Expected

[2023-09-23T10:26:55.531Z] 2023-09-23T10:26:54Z:     azure-arm.ubuntu:     <int>: 1

[2023-09-23T10:26:55.531Z] 2023-09-23T10:26:54Z:     azure-arm.ubuntu: to equal

[2023-09-23T10:26:55.531Z] 2023-09-23T10:26:54Z:     azure-arm.ubuntu:     <int>: 0

[2023-09-23T10:26:55.531Z] 2023-09-23T10:26:54Z:     azure-arm.ubuntu: Command: playwright_version: stdout: patterns not found: [1.37.1]

[2023-09-23T10:26:55.531Z] 2023-09-23T10:26:54Z:     azure-arm.ubuntu:

[2023-09-23T10:26:55.531Z] 2023-09-23T10:26:54Z:     azure-arm.ubuntu: Total Duration: 1.022s

[2023-09-23T10:26:55.531Z] 2023-09-23T10:26:54Z:     azure-arm.ubuntu: Count: 17, Failed: 2, Skipped: 0

[2023-09-23T10:26:55.531Z] 2023-09-23T10:26:54Z:     azure-arm.ubuntu: Retrying in 1s (elapsed/timeout time: 1.023s/5s)
```

I think we could solve this by tuning the call to goss to use jenkins user here https://github.com/jenkins-infra/packer-images/blob/5fd72a2d71e441600d2ed8ff05eb3af55e9e0df2/build-jenkins-agent-ubuntu.pkr.hcl#L48-L54

**Comment by hainenber** [comment]
Apology for the back-and-forth 😞 (it's quite a long thread already) but I've managed to make a change to `goss validate` command. It now runs under `jenkins` user. Let me know if it works out on VM. Thanks!

**Comment by smerle33** [comment]
> Apology for the back-and-forth 😞 (it's quite a long thread already) but I've managed to make a change to `goss validate` command. It now runs under `jenkins` user. Let me know if it works out on VM. Thanks!

do not apology, it's not as easy as it seems ... sorry not to have time and knowledge to help you more .. there is still an error  : 
```
2023-09-26T09:08:14Z: ==> docker.ubuntu: Provisioning with shell script: /tmp/packer-shell2957735042

[2023-09-26T09:08:14.288Z] 2023-09-26T09:08:14Z: ==> docker.ubuntu: + goss --version

[2023-09-26T09:08:14.288Z] 2023-09-26T09:08:14Z:     docker.ubuntu: goss version v0.4.2

[2023-09-26T09:08:14.288Z] 2023-09-26T09:08:14Z: ==> docker.ubuntu: + su - jenkins -c goss --gossfile /tmp/goss.yaml validate --retry-timeout 5s

[2023-09-26T09:08:15.652Z] 2023-09-26T09:08:15Z:     docker.ubuntu: .................FF

[2023-09-26T09:08:15.652Z] 2023-09-26T09:08:15Z:     docker.ubuntu:

[2023-09-26T09:08:15.652Z] 2023-09-26T09:08:15Z:     docker.ubuntu: Failures/Skipped:

[2023-09-26T09:08:15.652Z] 2023-09-26T09:08:15Z:     docker.ubuntu:

[2023-09-26T09:08:15.652Z] 2023-09-26T09:08:15Z:     docker.ubuntu: Command: playwright_version: exit-status:

[2023-09-26T09:08:15.652Z] 2023-09-26T09:08:15Z:     docker.ubuntu: Expected

[2023-09-26T09:08:15.652Z] 2023-09-26T09:08:15Z:     docker.ubuntu:     1

[2023-09-26T09:08:15.652Z] 2023-09-26T09:08:15Z:     docker.ubuntu: to be numerically eq

[2023-09-26T09:08:15.652Z] 2023-09-26T09:08:15Z:     docker.ubuntu:     0

[2023-09-26T09:08:15.652Z] 2023-09-26T09:08:15Z:     docker.ubuntu: Command: playwright_version: stdout:

[2023-09-26T09:08:15.652Z] 2023-09-26T09:08:15Z:     docker.ubuntu: Expected

[2023-09-26T09:08:15.652Z] 2023-09-26T09:08:15Z:     docker.ubuntu:     "object: *bytes.Reader"

[2023-09-26T09:08:15.652Z] 2023-09-26T09:08:15Z:     docker.ubuntu: to have patterns

[2023-09-26T09:08:15.652Z] 2023-09-26T09:08:15Z:     docker.ubuntu:     ["1.37.1"]

[2023-09-26T09:08:15.652Z] 2023-09-26T09:08:15Z:     docker.ubuntu: the missing elements were

[2023-09-26T09:08:15.652Z] 2023-09-26T09:08:15Z:     docker.ubuntu:     ["1.37.1"]

[2023-09-26T09:08:15.652Z] 2023-09-26T09:08:15Z:     docker.ubuntu:

[2023-09-26T09:08:15.652Z] 2023-09-26T09:08:15Z:     docker.ubuntu: Total Duration: 1.008s

[2023-09-26T09:08:15.652Z] 2023-09-26T09:08:15Z:     docker.ubuntu: Count: 19, Failed: 2, Skipped: 0

[2023-09-26T09:08:15.652Z] 2023-09-26T09:08:15Z:     docker.ubuntu: Retrying in 1s (elapsed/timeout time: 1.009s/5s)
```

**Comment by dduportal** [comment]
Hi @hainenber ! Since https://github.com/jenkins-infra/packer-images/pull/836, the `goss` test harness is run by the `jenkins` user : can you rebase your PR here and update it to fit to this new model?

File: build-jenkins-agent-ubuntu.pkr.hcl [hunk 3]
```diff
@@ -50,7 +50,7 @@ build {
     inline = [
       "source /home/jenkins/.asdf/asdf.sh", # Required as this is a non-interactive and non-login `bash`
       "goss --version",
-      "goss --gossfile /tmp/goss.yaml validate --retry-timeout 5s",
+      "su - jenkins -c 'goss --gossfile /tmp/goss.yaml validate --retry-timeout 5s'",
```
Comment (dduportal) [inline build-jenkins-agent-ubuntu.pkr.hcl:53]:
```suggestion
      "goss --gossfile /tmp/goss.yaml validate --retry-timeout 5s",
```

As the line 49 already specifies the user `jenkins` for execution

File: goss/goss.yaml [hunk 4]
```diff
@@ -38,6 +38,13 @@ command:
     exit-status: 0
     stdout:
       - 18.17.0
+    stderr: []
```
Comment (dduportal) [inline goss/goss.yaml:41]:
```suggestion
```

Not needed (nitpick)

**Review by dduportal — APPROVED** [review]
I'v rebase and push-forced to update and fix conflicts, as the PR looks good to me but there were goss errors in the last build.

I'm watching the build and playing with it locally but LGTM 👍 thanks!

File: goss/goss.yaml [hunk 5]
```diff
@@ -38,6 +38,11 @@ command:
     exit-status: 0
     stdout:
       - 18.18.2
+  playwright_version:
+    exec: "playwright --version'" # Playwright is installed as asdf shims for `jenkins` user only
```
Comment (dduportal) [inline goss/goss.yaml:42]:
```suggestion
    exec: playwright --version
```

There was a minor typo causing the goss failure (trailing simple quote `'`).

I'm merging the suggestion above (with removal of double quotes and trailing comment to make the comment easier to get)

**Review by dduportal — APPROVED** [review]
Tested locally and works very well!

**Comment by hainenber** [comment]
Thank you all for your support!

**Comment by dduportal** [comment]
@hainenber 👏 : https://github.com/jenkins-infra/packer-images/pull/885

**Pull request merged**